tinydnssec-1.05/dnscache-conf.8
.TH dnscache-conf 8

.SH NAME
dnscache-conf \- sets up a DNS cache service.

.SH SYNOPSIS
.B dnscache-conf 
.I acct
.I logacct
.I D
[
.I ip
]

.SH DESCRIPTION
This is a reference page.
For tutorial information, see the instructions for
.br
.B workstations
(http://cr.yp.to/djbdns/run-cache.html),
.br
.B home computers
(http://cr.yp.to/djbdns/run-cache-home.html),
.br
.B external caches
(http://cr.yp.to/djbdns/run-cache-x.html),
or
.br
.B upgrading from BIND
(http://cr.yp.to/djbdns/run-cache-bind-1.html).

.B dnscache-conf
creates a service directory 
.I D
that runs
.BR dnscache (8).
The name 
.I D
must start with a slash
and must not contain any special characters.
Normally 
.I D
is 
.IR /etc/indimail/dnscache .

You can run the service under
.BR svscan (8)
by creating a symbolic link in the 
.BR svscan (8)
directory:

ln -s 
.I D
/service

The service will start within five seconds,
and will be restarted upon reboot.
You can use
.BR svc (8)
to control the service.

.B dnscache-conf
arranges for 
.BR dnscache (8)
to chroot to 
.IR D /root
and to run under the uid and gid of 
.IR acct .
The name
.I acct
must not contain any special characters.

.B dnscache-conf
arranges for 
.BR dnscache (8)
to listen for UDP packets and TCP connections on port 53 of 
.IR ip .

.I ip
is optional;
if it is not supplied,
.B dnscache-conf
arranges for 
.BR dnscache (8)
to listen on 127.0.0.1.

.B dnscache-conf
creates 
.IR D /root/ip/127.0.0.1
so that 
.BR dnscache (8)
will accept queries from 127.0.0.1.

.B dnscache-conf
puts 128 bytes of not-particularly-secret data
into 
.IR D /seed
and arranges for 
.BR dnscache (8)
to pass 
.IR D /seed
to 
.BR dns_random_init (3).
If your system has a good source of random data,
you can replace 
.IR D /seed
with 128 bytes of data from that source.

.B dnscache-conf
creates an automatically rotated log directory in
.IR D /log/main.
The logs are owned by 
.IR logacct .
The corresponding 
.BR multilog (8)
processes run under the uid and gid of 
.IR logacct .
The name 
.I logacct
must not contain any special characters.

.SH SEE ALSO
dnscache(8),
svscan(8),
svc(8),
multilog(8),
axfrdns-conf(8),
rbldns-conf(8),
tinydns-conf(8),
dqcache-conf(8),
curvedns-conf(8),
pickdns-conf(8),
walldns-conf(8)

http://cr.yp.to/djbdns.html


tinydnssec-1.05/dnscache-conf.c
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "hasdevtcp.h"
#ifdef HASDEVTCP
#include <sys/mkdev.h>
#endif
#include <pwd.h>
#include "strerr.h"
#include "buffer.h"
#include "uint32.h"
#include "taia.h"
#include "str.h"
#include "open.h"
#include "error.h"
#include "exit.h"
#include "auto_home.h"
#include "generic-conf.h"

#define FATAL "dnscache-conf: fatal: "

void usage(void)
{
  strerr_die1x(100,"dnscache-conf: usage: dnscache-conf acct logacct /dnscache [ myip ]");
}

int fdrootservers;
char rootserversbuf[64];
buffer ssrootservers;

char *dir;
char *user;
char *loguser;
struct passwd *pw;
const char *myip;

uint32 seed[32];
int seedpos = 0;

void seed_adduint32(uint32 u)
{
  int i;

  seed[seedpos] += u;
  if (++seedpos == 32) {
    for (i = 0;i < 32;++i) {
      u = ((u ^ seed[i]) + 0x9e3779b9) ^ (u << 7) ^ (u >> 25);
      seed[i] = u;
    }
    seedpos = 0;
  }
}

void seed_addtime(void)
{
  struct taia t;
  char tpack[TAIA_PACK];
  int i;

  taia_now(&t);
  taia_pack(tpack,&t);
  for (i = 0;i < TAIA_PACK;++i)
    seed_adduint32(tpack[i]);
}

int main(int argc,char **argv)
{
  seed_addtime();
  seed_adduint32(getpid());
  seed_adduint32(getppid());
  seed_adduint32(getuid());
  seed_adduint32(getgid());

  user = argv[1];
  if (!user) usage();
  loguser = argv[2];
  if (!loguser) usage();
  dir = argv[3];
  if (!dir) usage();
  if (dir[0] != '/') usage();
  myip = argv[4];
  if (!myip) myip = "127.0.0.1";

  pw = getpwnam(loguser);
  seed_addtime();
  if (!pw)
    strerr_die3x(111,FATAL,"unknown account ",loguser);

  if (chdir(pw->pw_dir) == -1)
    strerr_die4sys(111,FATAL,"unable to switch to ",pw->pw_dir,": ");

  fdrootservers = open_read("dnsroots.local");
  if (fdrootservers == -1) {
    if (errno != error_noent)
      strerr_die2sys(111,FATAL,"unable to open dnsroots.local: ");
    fdrootservers = open_read("dnsroots.global");
    if (fdrootservers == -1)
      strerr_die2sys(111,FATAL,"unable to open dnsroots.global: ");
  }

  init(dir,FATAL);

  seed_addtime(); makedir("log");
  seed_addtime(); perm(02755);
  seed_addtime(); makedir("log/main");
  seed_addtime(); owner(pw->pw_uid,pw->pw_gid);
  seed_addtime(); perm(02755);
  seed_addtime(); start("log/status"); finish();
  seed_addtime(); owner(pw->pw_uid,pw->pw_gid);
  seed_addtime(); perm(0644);
  seed_addtime(); makedir("env");
  seed_addtime(); perm(02755);
  seed_addtime(); start("env/ROOT"); outs(dir); outs("/root\n"); finish();
  seed_addtime(); perm(0644);
  seed_addtime(); start("env/IP"); outs(myip); outs("\n"); finish();
  seed_addtime(); perm(0644);
  seed_addtime(); start("env/IPSEND"); outs("0.0.0.0\n"); finish();
  seed_addtime(); perm(0644);
  seed_addtime(); start("env/CACHESIZE"); outs("1000000\n"); finish();
  seed_addtime(); perm(0644);
  seed_addtime(); start("env/DATALIMIT"); outs("3000000\n"); finish();
  seed_addtime(); perm(0644);
  seed_addtime(); start("run");
  outs("#!/bin/sh\nexec 2>&1\nexec <seed\nexec envdir ./env sh -c '\n  exec envuidgid "); outs(user);
  outs(" softlimit -o250 -d \"$DATALIMIT\" ");
  outs(auto_home); outs("/bin/dnscache\n'\n"); finish();
  seed_addtime(); perm(0755);
  seed_addtime(); start("log/run");
  outs("#!/bin/sh\nexec setuidgid "); outs(loguser);
  outs(" multilog t ./main\n"); finish();
  seed_addtime(); perm(0755);
  seed_addtime(); makedir("root");
  seed_addtime(); perm(02755);
  seed_addtime(); makedir("root/ip");
  seed_addtime(); perm(02755);
  seed_addtime(); start("root/ip/127.0.0.1"); finish();
  seed_addtime(); perm(0600);
  seed_addtime(); makedir("root/servers");
  seed_addtime(); perm(02755);
  seed_addtime(); start("root/servers/@");
  buffer_init(&ssrootservers,buffer_unixread,fdrootservers,rootserversbuf,sizeof rootserversbuf);
  copyfrom(&ssrootservers);
  finish();
  seed_addtime(); perm(0644);
  seed_addtime();

  start("seed"); out((char *) seed,128); finish();
  perm(0600);

#ifdef HASDEVTCP
  makedir("root/etc");
  perm(02755);
  makedir("root/dev");
  perm(02755);
  start("root/etc/netconfig");
  outs("tcp tpi_cots_ord v inet tcp /dev/tcp -\n");
  outs("udp tpi_clts v inet udp /dev/udp -\n");
  finish();
  perm(0645);
  umask(000);
  if (mknod("root/dev/tcp",S_IFCHR | 0667,makedev(11,42)) == -1)
    strerr_die4sys(111,FATAL,"unable to create device ",dir,"/root/dev/tcp: ");
  if (mknod("root/dev/udp",S_IFCHR | 0667,makedev(11,41)) == -1)
    strerr_die4sys(111,FATAL,"unable to create device ",dir,"/root/dev/udp: ");
  umask(022);
#endif

  _exit(0);
}
